castle/cms/browser/login.py

```diff
@@ -187,6 +189,25 @@ def login(self):
             })
 
         try:
+            if hasattr(self.context,'portal_registry'):
+                backend_urls = self.context.portal_registry['plone.backend_url']
+                only_allow_login_to_backend_urls = self.context.portal_registry['plone.only_allow_login_to_backend_urls']
+                pstate = getMultiAdapter(
+                    (self.context, self.request),
+                    name=u'plone_portal_state'
+                )
+                portal_url = pstate.portal_url()
```

Comment:
This is the new standard way to get the portal_url: `from plone import api; url = api.portal.get().absolute_url()` from https://docs.plone.org/develop/plone.api/docs/portal.html#get-portal-url or this older and more common `from Products.CMFCore.utils import getToolByName; portal_url = getToolByName(presenter, 'portal_url')` from https://docs.plone.org/external/plone.app.dexterity/docs/testing/mock-testing.html and https://docs.plone.org/develop/plone/misc/context.html#gettoolbyname